File: R/twliConsolidate/R/consolidate.R
.SUPPORTED_AGGS <- c("sum", "mean", "first", "last", "min", "max",
                     "weighted_mean", "recompute")

#' Aggregate a panel dataset to Stable Analysis Units (SAU).
#'
#' Mirrors the algorithm of the Python ``twli_consolidate.consolidate_panel``.
#'
#' @param df A long-format panel data.frame with one row per (V_ID, year).
#' @param vid_col Name of the village-code column.
#' @param year_col Name of the year column.
#' @param vars_spec Named list of per-variable aggregation rules. Each value
#'   may be a single character (``"sum"``, ``"mean"``, etc.) or a list with
#'   element ``agg`` and (for ``weighted_mean``) ``weight`` or
#'   (for ``recompute``) ``expr``.
#' @param lineage A ``twliLineage`` object from [build_lineage()].
#' @param extra_group_cols Optional character vector of extra columns to keep
#'   (first value within each SAU).
#' @param keep_member_list Logical; if TRUE, append a ``members`` column with
#'   the original V_IDs aggregated for each (SAU, year) row.
#' @return A data.frame with columns
#'   ``sau_id, <year_col>, <extra cols>, <vars>, members?``.
#' @export
consolidate_panel <- function(df,
                              vid_col,
                              year_col,
                              vars_spec,
                              lineage,
                              extra_group_cols = NULL,
                              keep_member_list = TRUE) {
  if (!vid_col %in% names(df))  stop(sprintf("vid_col '%s' not in dataframe", vid_col))
  if (!year_col %in% names(df)) stop(sprintf("year_col '%s' not in dataframe", year_col))
  spec <- .normalize_spec(vars_spec)
  .validate_columns(df, spec)

  work <- df
  work$`__sau__` <- sau_of(lineage, work[[vid_col]])

  # Phase 1: per-row pre-multiplication for weighted_mean
  aux <- list()
  for (col in names(spec)) {
    rule <- spec[[col]]
    if (rule$agg == "weighted_mean") {
      num <- paste0("__num__", col)
      den <- paste0("__den__", col)
      work[[num]] <- work[[col]] * work[[rule$weight]]
      work[[den]] <- work[[rule$weight]]
      aux[[col]] <- list(num = num, den = den)
    }
  }

  # Phase 2: group + summarise
  group_keys <- c("__sau__", year_col)
  agg_exprs <- list()
  for (col in names(spec)) {
    a <- spec[[col]]$agg
    if (a == "sum")          agg_exprs[[col]] <- rlang::expr(sum(!!rlang::sym(col), na.rm = TRUE))
    else if (a == "mean")    agg_exprs[[col]] <- rlang::expr(mean(!!rlang::sym(col), na.rm = TRUE))
    else if (a == "first")   agg_exprs[[col]] <- rlang::expr(dplyr::first(!!rlang::sym(col)))
    else if (a == "last")    agg_exprs[[col]] <- rlang::expr(dplyr::last(!!rlang::sym(col)))
    else if (a == "min")     agg_exprs[[col]] <- rlang::expr(min(!!rlang::sym(col), na.rm = TRUE))
    else if (a == "max")     agg_exprs[[col]] <- rlang::expr(max(!!rlang::sym(col), na.rm = TRUE))
    else if (a == "weighted_mean") {
      num <- aux[[col]]$num; den <- aux[[col]]$den
      agg_exprs[[num]] <- rlang::expr(sum(!!rlang::sym(num), na.rm = TRUE))
      agg_exprs[[den]] <- rlang::expr(sum(!!rlang::sym(den), na.rm = TRUE))
    }
    # recompute: handled in phase 3
  }
  if (length(extra_group_cols)) {
    for (c in extra_group_cols) {
      if (c %in% names(df) && !c %in% names(agg_exprs)) {
        agg_exprs[[c]] <- rlang::expr(dplyr::first(!!rlang::sym(c)))
      }
    }
  }

  grouped <- work %>%
    dplyr::group_by(dplyr::across(dplyr::all_of(group_keys))) %>%
    dplyr::summarise(!!!agg_exprs, .groups = "drop")

  # Phase 3: finalise weighted_mean and recompute
  for (col in names(spec)) {
    rule <- spec[[col]]
    if (rule$agg == "weighted_mean") {
      num <- aux[[col]]$num; den <- aux[[col]]$den
      grouped[[col]] <- ifelse(grouped[[den]] == 0, NA_real_,
                               grouped[[num]] / grouped[[den]])
      grouped[[num]] <- NULL; grouped[[den]] <- NULL
    }
  }
  for (col in names(spec)) {
    rule <- spec[[col]]
    if (rule$agg == "recompute") {
      e <- rlang::parse_expr(rule$expr)
      grouped[[col]] <- rlang::eval_tidy(e, data = grouped)
    }
  }

  grouped <- dplyr::rename(grouped, sau_id = `__sau__`)

  if (keep_member_list) {
    members_df <- work %>%
      dplyr::group_by(dplyr::across(dplyr::all_of(group_keys))) %>%
      dplyr::summarise(members = list(sort(unique(.data[[vid_col]]))),
                       .groups = "drop") %>%
      dplyr::rename(sau_id = `__sau__`)
    grouped <- dplyr::left_join(grouped, members_df, by = c("sau_id", year_col))
  }

  var_order <- names(spec)
  front <- c("sau_id", year_col)
  if (length(extra_group_cols)) {
    front <- c(front, intersect(extra_group_cols, names(grouped)))
  }
  tail <- if (keep_member_list) "members" else character(0)
  cols_final <- c(front, intersect(var_order, names(grouped)), tail)
  grouped <- grouped[, cols_final, drop = FALSE]
  dplyr::arrange(grouped, sau_id, .data[[year_col]])
}

.normalize_spec <- function(vars_spec) {
  out <- list()
  for (col in names(vars_spec)) {
    rule <- vars_spec[[col]]
    if (is.character(rule) && length(rule) == 1L) rule <- list(agg = rule)
    if (is.null(rule$agg)) stop(sprintf("vars_spec['%s'] missing 'agg'", col))
    if (!rule$agg %in% .SUPPORTED_AGGS) {
      stop(sprintf("vars_spec['%s'].agg = '%s' must be one of: %s",
                   col, rule$agg, paste(.SUPPORTED_AGGS, collapse = ", ")))
    }
    if (rule$agg == "weighted_mean" && is.null(rule$weight)) {
      stop(sprintf("vars_spec['%s'] weighted_mean needs 'weight'", col))
    }
    if (rule$agg == "recompute" && is.null(rule$expr)) {
      stop(sprintf("vars_spec['%s'] recompute needs 'expr'", col))
    }
    out[[col]] <- rule
  }
  out
}

.validate_columns <- function(df, spec) {
  for (col in names(spec)) {
    rule <- spec[[col]]
    if (rule$agg == "recompute") next
    if (!col %in% names(df)) stop(sprintf("variable '%s' not in dataframe", col))
    if (rule$agg == "weighted_mean" && !rule$weight %in% names(df)) {
      stop(sprintf("weight column '%s' (for %s) not in dataframe",
                   rule$weight, col))
    }
  }
}

File: python/twli_consolidate/consolidate.py
"""Aggregate panel data to Stable Analysis Units."""
from __future__ import annotations

from typing import Any, Mapping

import pandas as pd

from .lineage import Lineage


SUPPORTED_AGGS = {"sum", "mean", "first", "last", "min", "max",
                  "weighted_mean", "recompute", "keep_if_unique"}


def _keep_if_unique(s: pd.Series):
    """Return the single value if the group has exactly one V_ID member,
    otherwise NaN.

    Use case: 衍生統計量（中位數、分位數、標準差、變異係數等）無法從區級彙總
    重算。對 SAU 內單一 V_ID（未整併）保留原值；對多 V_ID SAU（整併過）標 NaN。
    """
    s = s.dropna()
    return s.iloc[0] if len(s) == 1 else float("nan")


def consolidate_panel(
    df: pd.DataFrame,
    *,
    vid_col: str,
    year_col: str,
    vars_spec: Mapping[str, Mapping[str, Any] | str],
    lineage: Lineage,
    extra_group_cols: list[str] | None = None,
    keep_member_list: bool = True,
) -> pd.DataFrame:
    """Recompute a panel dataset over Stable Analysis Units.

    Parameters
    ----------
    df : DataFrame
        Long-format panel. Must contain ``vid_col`` and ``year_col`` plus any
        variable columns referenced in ``vars_spec``.
    vid_col : str
        Name of the village-code column (matches V_IDs in the crosswalk).
    year_col : str
        Name of the year column (Gregorian integer).
    vars_spec : mapping
        Per-variable aggregation rule. Either a string shorthand
        (``"sum" | "mean" | "first" | "last" | "min" | "max" | "keep_if_unique"``)
        or a dict::

            {"agg": "sum"}
            {"agg": "weighted_mean", "weight": "P_CNT"}
            {"agg": "recompute", "expr": "P_CNT / AREA"}
            {"agg": "keep_if_unique"}  # 對 SAU 內單一 V_ID 保留原值，多 V_ID 標 NaN

        ``recompute`` rules run **after** all other aggregations. The
        expression is evaluated with :py:meth:`pandas.DataFrame.eval` and may
        reference any already-aggregated column.
    lineage : Lineage
        Built from :func:`twli_consolidate.build_lineage`.
    extra_group_cols : list[str], optional
        Additional columns to keep in grouping (e.g. ``["COUNTY", "TOWN"]``).
        Within an SAU these may differ across members; the value of the first
        member encountered is used.
    keep_member_list : bool, default True
        If True, include a ``members`` column listing all original V_IDs that
        were aggregated for that (sau_id, year) row.

    Returns
    -------
    DataFrame with one row per (sau_id, year), columns:
    ``[sau_id, year, *extra_group_cols, *aggregated vars, members?]``
    """
    if vid_col not in df.columns:
        raise KeyError(f"vid_col {vid_col!r} not in dataframe")
    if year_col not in df.columns:
        raise KeyError(f"year_col {year_col!r} not in dataframe")

    spec = _normalize_spec(vars_spec)
    _validate_columns(df, spec)

    work = df.copy()
    work["__sau__"] = work[vid_col].map(lineage.sau_of)

    # Phase 1: per-row pre-multiplication for weighted_mean
    aux_cols: dict[str, str] = {}
    for col, rule in spec.items():
        if rule["agg"] == "weighted_mean":
            wcol = rule["weight"]
            num = f"__num__{col}"
            den = f"__den__{col}"
            work[num] = work[col] * work[wcol]
            work[den] = work[wcol]
            aux_cols[col] = (num, den)

    # Phase 2: build groupby aggregation map
    agg_map: dict[str, Any] = {}
    for col, rule in spec.items():
        a = rule["agg"]
        if a in {"sum", "mean", "first", "last", "min", "max"}:
            agg_map[col] = a
        elif a == "keep_if_unique":
            agg_map[col] = _keep_if_unique
        elif a == "weighted_mean":
            num, den = aux_cols[col]
            agg_map[num] = "sum"
            agg_map[den] = "sum"
        elif a == "recompute":
            pass  # handled in phase 3
    if extra_group_cols:
        for c in extra_group_cols:
            if c in df.columns:
                agg_map.setdefault(c, "first")

    grouped = work.groupby(["__sau__", year_col], as_index=False, dropna=False).agg(agg_map)

    # Phase 3: finalize weighted_mean and recompute
    for col, rule in spec.items():
        if rule["agg"] == "weighted_mean":
            num, den = aux_cols[col]
            grouped[col] = grouped[num] / grouped[den].where(grouped[den] != 0)
            grouped = grouped.drop(columns=[num, den])

    for col, rule in spec.items():
        if rule["agg"] == "recompute":
            grouped[col] = grouped.eval(rule["expr"])

    grouped = grouped.rename(columns={"__sau__": "sau_id"})

    if keep_member_list:
        members = (
            work.groupby(["__sau__", year_col], dropna=False)[vid_col]
            .agg(lambda s: sorted(set(s)))
            .reset_index()
            .rename(columns={"__sau__": "sau_id", vid_col: "members"})
        )
        grouped = grouped.merge(members, on=["sau_id", year_col], how="left")

    # Order: sau_id, year, extras, vars, members
    var_order = list(spec.keys())
    front = ["sau_id", year_col]
    if extra_group_cols:
        front += [c for c in extra_group_cols if c in grouped.columns]
    tail = ["members"] if keep_member_list else []
    cols_final = front + [c for c in var_order if c in grouped.columns] + tail
    return grouped[cols_final].sort_values(["sau_id", year_col]).reset_index(drop=True)


def _normalize_spec(vars_spec) -> dict[str, dict[str, Any]]:
    out: dict[str, dict[str, Any]] = {}
    for col, rule in vars_spec.items():
        if isinstance(rule, str):
            rule = {"agg": rule}
        if "agg" not in rule:
            raise ValueError(f"vars_spec[{col!r}] missing 'agg' key")
        if rule["agg"] not in SUPPORTED_AGGS:
            raise ValueError(
                f"vars_spec[{col!r}].agg = {rule['agg']!r}; "
                f"must be one of {sorted(SUPPORTED_AGGS)}"
            )
        if rule["agg"] == "weighted_mean" and "weight" not in rule:
            raise ValueError(f"vars_spec[{col!r}] weighted_mean needs 'weight'")
        if rule["agg"] == "recompute" and "expr" not in rule:
            raise ValueError(f"vars_spec[{col!r}] recompute needs 'expr'")
        out[col] = dict(rule)
    return out


def _validate_columns(df: pd.DataFrame, spec: dict) -> None:
    for col, rule in spec.items():
        if rule["agg"] == "recompute":
            continue
        if col not in df.columns:
            raise KeyError(f"variable {col!r} not in dataframe")
        if rule["agg"] == "weighted_mean" and rule["weight"] not in df.columns:
            raise KeyError(f"weight column {rule['weight']!r} (for {col}) not in dataframe")
